### DISCUSSION TITLE
Which turn function is more accurate and reliable?

### DISCUSSION THREAD

**Discussion by Samb19802000** [pybricks/support#1263] [General]
Hi Experts,

We can use the following two methods to turn the robot to a target angle.

1. DriveBase.drive(0, turn_rate: xx) with hub.imu.heading()
2. DriveBase.turn(xx)

my question is which method is more reliable and accurate? 

I appreciate your feedback.

Thanks!!

**Comment by laurensvalk** [comment]
The accuracy is the same but they serve different purposes. If you just want to turn by a given angle, just use `turn`.

`drive` is more useful for things like line-following where you constantly change the turn rate based on the light reading, for example.

**Reply by Samb19802000** [reply]
Thanks for the reply. 
I use the following code to turn left, and the turn is not smooth. As an example, if wanted to take a 90deg turn, I expected to turn ~88deg and then correct the delta. Instead, it takes 2,3 turns to complete the turn. 


```
def left_turn_angle(angle):
    angle = (-angle) + hub.imu.heading()
    delta = angle - hub.imu.heading()
    while round(delta) not in range(-TOLERANCE, TOLERANCE) :
        drive_base.turn(delta)
        delta = angle - hub.imu.heading()

left_turn_angle(90)
```

Thanks,

**Reply by dlech** [reply]
Depending on how big `TOLERANCE` is, ` not in range(-TOLERANCE, TOLERANCE)` could take a long time because it will check every single integer value between `-TOLERANCE` and `TOLERANCE`!

Perhaps better would be:

```python
while not (-TOLERANCE <= delta <= TOLERANCE):
```

But looking at the bigger picture, it looks like you are trying to define a new function that does what `drive_base.turn()` already does if you just call it by itself. Pybricks Beta has a `use_gyro` option for the `DriveBase` constructor to tell it to use the gyro to assist with control.

**Reply by Samb19802000** [reply]
Thanks,
my tolerance is 1deg. Looks like I reinvented the wheel :-). 
Did you refer to `GyroDriveBase`

**Reply by dlech** [reply]
Yes. Although we removed `GyroDriveBase` and replaced it with a `use_gyro=True` option for the `DriveBase` in v3.3.0b8.

**Reply by Samb19802000** [reply]
Thanks!!

**Comment by laurensvalk** [comment]
Reading between the lines, I think you're asking about a form of absolute angles.

To accomplish that, you can calculate the _remaining_ angle and combine it with @dlech's answer:


```python

drive_base.use_gyro(True)

# For example if you want to go to 180
remaining_angle = 180 - hub.imu.heading()
drive_base.turn(remaining_angle)
```

**Reply by Samb19802000** [reply]
Thanks!!
